src/sb16.js

```diff
@@ -1169,19 +1218,196 @@ function register_mixer_write(address, handler)
+
+// Master Volume Left.
+register_mixer_volume(0x30, "master", "left");
+// Master Volume Right.
+register_mixer_volume(0x31, "master", "right");
+// Voice Volume Left.
+register_mixer_volume(0x32, "dac", "left");
+// Voice Volume Right.
+register_mixer_volume(0x33, "dac", "right");
+// MIDI Volume Left. TODO.
+//register_mixer_volume(0x34, "synth", "left");
+// MIDI Volume Right. TODO.
+//register_mixer_volume(0x35, "synth", "right);
+// CD Volume Left. TODO.
+//register_mixer_volume(0x36, "cd", "left);
+// CD Volume Right. TODO.
+//register_mixer_volume(0x37, "cd", "right);
+// Line Volume Left. TODO.
+//register_mixer_volume(0x38, "line", "left");
+// Line Volume Right. TODO.
+//register_mixer_volume(0x39, "line", "right");
+// Mic Volume. TODO.
+//register_mixer_volume(0x3A, "mic", "both");
+
+// PC Speaker Volume.
+register_mixer_read(0x3B);
+register_mixer_write(0x3B, function(data)
+{
+    this.mixer_registers[0x3B] = data;
+    this.bus.send("mixer-volume", ["pcspeaker", "both", (data >>> 6) * 6 - 18]);
+});
+
+// Output Mixer Switches. TODO.
+//register_mixer_read(0x3C);
+//register_mixer_write(0x3C, function(data)
+//{
+//    this.mixer_registers[0x3C] = data;
+//
+//    if(data & 0x01) this.bus.send("mixer-connect", ["mic", "both"]);
+//    else this.bus.send("mixer-disconnect", ["mic", "both"]);
+//
+//    if(data & 0x02) this.bus.send("mixer-connect", ["cd", "right"]);
+//    else this.bus.send("mixer-disconnect", ["cd", "right"]);
+//
+//    if(data & 0x04) this.bus.send("mixer-connect", ["cd", "left"]);
+//    else this.bus.send("mixer-disconnect", ["cd", "left"]);
+//
+//    if(data & 0x08) this.bus.send("mixer-connect", ["line", "right"]);
+//    else this.bus.send("mixer-disconnect", ["line", "right"]);
+//
+//    if(data & 0x10) this.bus.send("mixer-connect", ["line", "left"]);
+//    else this.bus.send("mixer-disconnect", ["line", "left"]);
+//});
+
+// Input Mixer Left Switches. TODO.
+//register_mixer_read(0x3D);
+//register_mixer_write(0x3D);
+
+// Input Mixer Right Switches. TODO.
+//register_mixer_read(0x3E);
+//register_mixer_write(0x3E);
+
+// Input Gain Left. TODO.
+//register_mixer_read(0x3F);
+//register_mixer_write(0x3F);
+
+// Input Gain Right. TODO.
+//register_mixer_read(0x40);
+//register_mixer_write(0x40);
+
+// Output Gain Left.
+register_mixer_read(0x41);
+register_mixer_write(0x41, function(data)
+{
+    this.mixer_registers[0x41] = data;
+    this.bus.send("mixer-gain-left", (data >>> 6) * 6);
+});
+
+// Output Gain Right.
+register_mixer_read(0x42);
+register_mixer_write(0x42, function(data)
+{
+    this.mixer_registers[0x42] = data;
+    this.bus.send("mixer-gain-right", (data >>> 6) * 6);
+});
+
+// Mic AGC. TODO.
+//register_mixer_read(0x43);
+//register_mixer_write(0x43);
+
+// Treble Left.
+register_mixer_read(0x44);
+register_mixer_write(0x44, function(data)
+{
+    this.mixer_registers[0x44] = data;
+    this.bus.send("mixer-treble-left", (data >>> 3) - (data >>> 7) - 14);
```

Comment:
Out of curiosity, what does this formula calculate?

Reply:
Ah this one's interesting. The manual says that the higher 4 bits are treated as a number, and this number's value gets mapped in the following way:

- Values 0 to 7 --> levels -14 to 0 dB in steps of 2 dB.
- Values from 8 to 15 --> levels 0 to 14 dB in steps of 2 dB. 

The middle two values, 7 and 8, both gets mapped to 0 dB. This elegant/hackish code does this using `data >>> 7`, which offsets the dB level by -2 dB when bit 7 is set.

Do you reckon I should make this clearer?

Reply:
It doesn't seem to yield the expected result:

```
In [1]: [(data << 4 >> 3) - (data << 4 >> 7) - 14 for data in range(16)]
Out[1]: [-14, -12, -10, -8, -6, -4, -2, 0, 1, 3, 5, 7, 9, 11, 13, 15]
```

I don't mind these bithacks (I have written many myself), but indeed a ternary might be clearer in this case.